planning/planning_scene_monitor/src/planning_scene_monitor.cpp

```diff
@@ -163,6 +165,11 @@ void planning_scene_monitor::PlanningSceneMonitor::initialize(const planning_sce
   moveit::tools::Profiler::ScopedStart prof_start;
   moveit::tools::Profiler::ScopedBlock prof_block("PlanningSceneMonitor::initialize");
 
+  // start our own spinner listening on our own callback_queue to become independent of any global callback queue
+  root_nh_.setCallbackQueue(&callback_queue_);
+  spinner_.reset(new ros::AsyncSpinner(1 /* threads */, &callback_queue_));
+  spinner_->start();
```

Comment:
On Wed, Aug 10, 2016 at 08:59:31PM -0700, Alain Sanguinetti wrote:
> I think this is probably a nice fix.

I see two problems with this:

- It should definitely be optional(enabled by default) because some people *do not want* to have unmanaged threads running on their systems.
  See your very own request https://github.com/ros-planning/moveit_ros/pull/701 for a recent example.

- For unclear reasons it has been explicitly disallowed to start ASyncSpinner's for different threads. See https://github.com/ros/ros_comm/issues/277
  The commits in question are definitely faulty, but we are unsure about the details at the moment.
  As a result this approach will break any code that instantiates a PlanningSceneMonitor in one thread and an(other) `AsyncSpinner` in a different thread.
  One example for this is RViz.
  
  An ugly workaround to this, that had been used in other places in MoveIt, is to require (without checking) a running `AsyncSpinner` in the background
  when certain functions are called. See https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group.h#L619
  If the bug in ros_comm would get fixed (also in indigo), we could remove that workaround.

> Maybe we should allow the user to specify how many threads he wants to use.

I would prefer to have it fully optional. If, as an alternative, the user could specify a `NodeHandle` for which he guarantees
that it will receive asyncronous updates this would include your use-case.